### ISSUE TITLE
Provide a visual selection feedback

### ISSUE THREAD

**Issue by angeliski** [olifolkerd/tabulator#2229]
**Is your feature request related to a problem? Please describe.**
The current select row module is nice, but I would like to have a visual feedback for the select function, maybe provided by a first column.
I think that is good to provide a better UX, the [Usability Heuristics](https://www.nngroup.com/articles/ten-usability-heuristics/) have a rule called **Visibility of system status**. When I first see the table I don't know if the table is selectable or not.

**Describe the solution you'd like**
The Tabulator provide a column feedback for that. Included a Select All Header 

Ex:
![image](https://user-images.githubusercontent.com/1574240/61672822-8581fe80-acc3-11e9-9f29-77ae31eb66ec.png)

**Describe alternatives you've considered**
I can build that behavior external of Tabulator core, but is very complex control that without using Tabulator core (using just callbacks and properties). And I belive is better to Tabulator have that feature.

**Additional context**
Maybe the issue #2208 is related

**Comment by olifolkerd** [comment]
Hey @angeliski 

This is already on the roadmap for the 4.5 release.

Cheers

Oli :)

**Comment by angeliski** [comment]
Nice @olifolkerd I can help? Maybe you can draft your design code ideia and I can help to put that in 4.4 release, would be nice :)

**Comment by olifolkerd** [comment]
Hey @angeliski 

Thanks for the offer, that would be great!

So this will involve tinkering around in Tabulators internal module structure, so what i would like to do is work through it with you piece by piece so i can introduce you to how it all works internally.

### Making a Pull Request
Here is a top level set of rules for pull requests:

-Your changes should consider all users, not just your specific usage case, I will not accept changes that reduce functionality in one area just to fix a different usage case

-Internal functionality changes should call internal functions, the external API functions are for end users

-Please try and match Tabulators current coding style

-Changes must be made to files in the src directory, the dist files are auto generated and changes there will be automatically overwritten on the next commit

-querySelectors are banned, tabulator works by storing references to elements and passing them arround, querySelectors are very inefficient and will result in automatic rejection of the pull request.


### The Tasks
So now thats out of the way, how to build the row selection features.

so there are three bits of functionality you will need to add:
- Some updates to the select_row module to accommodate the new functionality
- A formatter to show the tickbox
- A formatter to handle the header tickbox

What I would like you to start on is the first two in the list, making a formatter that is bound to the select state of the row. once that is done can you submit a pull request and i can talk you through the third part. This will need to work in a very specific way to be efficient so please do not even start the third task until you have made a pull request on the first two and we have outlined the next part of the plan.

you will need the ***gulp*** task runner installed to compile code changes, this can be installed with the command:
```
npm install -g gulp
```

you will then need to run ***gulp watch*** , this will watch your code for changes as you save them and compile the updated dist files

### Phase 1
Phase one will be to make a formatter that adds a tick box that can be used to toggle the selection state, and whos initial state is bound to the rows currently selected state, this can be done in the formater by getting the row component from the cell:
```js
var row = cell.getRow();
```

and then calling the ***isSelected*** function on the row. you can then select and deselect the row by calling the ***select*** and ***deselect*** function when the user click the select box.

You should call this formatter ***rowSelection*** and add it to the bottom of the ***Format.prototype.formatters*** object in '/src/js/modules/format.js`

### Phase 2
We now need to link this bar to the select_row module so that if the row selection is called programatically, the checkbox will still update without needing to trigger a reformat of the whole table.

To do this we are going to access some internal functions of Tabulator that should not be publicly use, *(ooooh exciting i know....)*

To start with you will need to make some changes to the select_row module, which can be found at '/src/js/modules/select_row.js'

In here you will need to create a new function on the prototype, lets call it ***registerRowSelectChecbox*** and it will take two arguments the row component of the row being registered and the element for the checkbox.

inside this function you will need to set the following:
```js
row._row.modules.select.checkboxEl = element;
```
this gets the Tabulator internal row object from the row component, then findes the modules object on it ***(this is where modules can safely store special props)***, you will need to check if the ***select*** property is set on it first and set it to an empty object if it isnt. then set the ***checkboxEl*** property on it to the checkbox element you have passed into the function.

This has now bound the checkbox element to the internal row object so we can use it later down the line if we need to.

Now you will need to update the ***_selectRow*** and ***deselectRows*** to update the checked state of the checkbox.

in both of these functions you should see a line like this:
```js
row.modules.select.selected = true; //or false depending on the function
```
under this you can then either check or uncheck the checkbox with the following line of code:
row.modules.select.checkboxEl.checked = true; //or false depending on what your are doing

And thats it, you should now have a special formatter that creates a checkbox that is automatically updated when the row selection changes (even when that is someone calling the ***selectRows*** function on the table.

When you have that up and running submit a pull request and i can take you through a bit more magic to get the global header tickbox working.

Cheers

Oli :)

**Comment by angeliski** [comment]
Nice @olifolkerd 
Some observations:
- I can use a SVG to control like the **tickCross** formatter?
-  That feature will be avaliable only when create a column using that formatter? I my mind that feature will work on put the selectable option true, what do you think?
- I can try put the #2054 togheter. You prefer separet that two issues?

**Comment by olifolkerd** [comment]
Hey @angeliski 

Thanks for getting back to me,

In answer to your questions:

- I would prefer if your use an input element type of checkbox, it will be what most users expect, there have infact be several users asking for it as an input checkbox already *(obviously once we are finished you can then make yourself a custom formatter to work how you want as you will understand how to do it, which is another good reason why it should be a formatter)*, later down the line we may update the formatter to allow for customization of this through the formatterParams, but i would prefer if we can just stay focused on getting the core functionality in for now and adding additional features after that

- Nope it must be available as a formatter, the whole point of Tabulator is to allow people to configure the table how they want, most people wont want a checkbox, others may want it on the right hand side of the table (think rtl languages), this way it makes it easy for people to have the table their way. i will include a demo on the site so that people understand how to use it, and it isnt much more code. One of the key points to take on board when writing for tabulator is that every function must be as customization as possible and work for all use cases not just one.

- Can we treat the CI tests as a totally separate issue as it is unrelated to this one. that way it will make it easier for us to finish this and get it merged in before moving onto the CI tests, otherwise we risk delaying them both and getting neither in

- if you create an `/examples` folder in the root of your project you can put any test html you like in there and reference the dist folder. there is a git ignore on the examples folder for just this purpose.

Let me know if you have any further questions

Cheers

Oli :)

**Comment by angeliski** [comment]
Thanks @olifolkerd Your points make sense.
I will do that soon as possible and open a PR to we work together.

**Comment by olifolkerd** [comment]
I eagerly await the PR :)

**Comment by olifolkerd** [comment]
Closed as it is now covered by the PR
